In [14]:
import pandas as pd 
import numpy as np
import seaborn as sns


## Exploration of Data Sets

### Exploring Scraped_Data.csv 

In [7]:
df_scraped_data_raw = pd.read_csv('scraped_data/scraped_data.csv')

In [8]:
df_scraped_data_raw.head(2)

scraping_id scraped_date        date  price  available
0            2   2018-03-15  2018-04-01    250          1
1            2   2018-03-15  2018-04-02    115          1

In [13]:
df_scraped_data_raw.describe()

scraping_id         price     available
count  1.630018e+07  1.630018e+07  1.630018e+07
mean   8.859516e+06  3.068724e+02  4.880770e-01
std    1.058208e+07  8.637483e+02  4.998578e-01
min    2.000000e+00  1.000000e+01  0.000000e+00
25%    3.835000e+03  1.000000e+02  0.000000e+00
50%    1.293900e+04  1.590000e+02  0.000000e+00
75%    2.147308e+07  3.000000e+02  1.000000e+00
max    2.470754e+07  2.500000e+04  1.000000e+00

sns.pairplot(df_scraped_data_raw)

### Exploring Scraped Listings 

In [237]:
df_scraped_listings_raw = pd.read_csv('scraped_data/scraped_listings.csv', encoding="ISO-8859-1")

In [238]:
df_scraped_listings_raw.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8445 entries, 0 to 8444
Data columns (total 14 columns):
scraping_id        8445 non-null int64
listing            8445 non-null object
city               8445 non-null object
lon                8445 non-null float64
lat                8445 non-null float64
mapped_location    8445 non-null object
name               8445 non-null object
capacity           8445 non-null int64
bathrooms          8445 non-null object
bedrooms           8445 non-null object
has_pool           8445 non-null int64
cleaning_fee       8445 non-null object
is_superhost       8445 non-null int64
host_name          8445 non-null object
dtypes: float64(2), int64(4), object(8)
memory usage: 923.8+ KB


In [239]:
df_scraped_listings_raw.head(1)

scraping_id                                listing        city         lon  \
0            2  https://www.airbnb.com/rooms/17720788  Scottsdale -111.916935   

         lat                                    mapped_location  \
0  33.496566  https://maps.google.com/maps?q=33.496565678911...   

                                                name  capacity bathrooms  \
0  Old Town Scottsdale Vacation Home & GIANTS Rental         7       2.5   

  bedrooms  has_pool cleaning_fee  is_superhost host_name  
0        2         1          150             1        Mj

In [240]:
df_scraped_listings_raw.shape

(8445, 14)

#### Exploring the Cities Column

In [241]:
cities = df_scraped_listings_raw['city'].unique()
sorted(cities)

['??',
 'Anthem',
 'Apache Junction',
 'Avondale',
 'Black Canyon City',
 'Buckeye',
 'Carefree',
 'Casa Grande',
 'Cave Creek',
 'Chandler',
 'Coolidge',
 'Downtown Phoenix',
 'El Mirage',
 'Flagstaff',
 'Florence',
 'Fort McDowell',
 'Fountain Hills',
 'Gilbert',
 'Glendale',
 'Gold Canyon',
 'Goodyear',
 'LAVEEN',
 'Laveen',
 'Litchfield Park',
 'Maricopa',
 'Maricopa County',
 'Mesa',
 'Morristown',
 'New River',
 'New River ',
 'Old Town Scottsdale ',
 'Paradise Valley',
 'Peoria',
 'Phoenix',
 'Phoenix ',
 'Phoenix  ',
 'Queen Creek',
 'Rio Verde',
 'San Tan Valley',
 'Scottsdale',
 'Scottsdale ',
 'Sun City',
 'Sun City West',
 'Sun Lakes',
 'Surprise',
 'Surprise, Arizona, US',
 'Tempe',
 'Tolleson',
 'Waddell',
 'Wittmann',
 'Youngtown',
 'phoenix',
 'scottsdale']

Note: from initial inspection, there seems to be city with ?? which might indicate an error. A few columns have lower case words and extra spacings. To solve this, we'll classify a few different error types and recategorize the listings city column based on this logic.

    1) City names should all be lowercased and stripped of spaces. Each listing in the df_listings dataframe needs to then be updated. We'll create some functions to do that. 
    2) Some of the listings might include US by mistake (e.g Surprise, Arizona, US). For future use, we should maintain a dictionary of proper listings and remove any text that doesn't match a city in the USA (if this is the market). 

In [242]:
df_scraped_listings_raw['city'].unique()

array(['Scottsdale', 'Phoenix', 'Tempe', 'Paradise Valley', 'Mesa',
       'Gilbert', 'Glendale', 'scottsdale', 'Maricopa County', 'Chandler',
       'Fountain Hills', 'Peoria', 'Cave Creek', 'Tolleson', 'Avondale',
       'Litchfield Park', 'Laveen', 'San Tan Valley', 'Surprise',
       'Queen Creek', 'Goodyear', 'LAVEEN', 'Sun City', 'Sun City West',
       'Rio Verde', 'Maricopa', 'Anthem', 'Gold Canyon', 'New River',
       'Apache Junction', 'Sun Lakes', 'Carefree', 'Old Town Scottsdale ',
       'El Mirage', 'Phoenix  ', 'Buckeye', 'Waddell', 'Casa Grande',
       'New River ', 'Morristown', 'Scottsdale ', 'Fort McDowell',
       'Flagstaff', 'Phoenix ', 'Downtown Phoenix', 'Florence', 'phoenix',
       'Surprise, Arizona, US', '??', 'Black Canyon City', 'Wittmann',
       'Coolidge', 'Youngtown'], dtype=object)

df_scraped_listings_raw['city'] = df_scraped_listings_raw['city'].str.lower().str.strip()

In [245]:
def fix_str_col(df, col):
    #get rid of edge case where the user added more than just city 
    df.loc[df[col] == 'Surprise, Arizona, US', col] = 'Surprise'
    
    #normalize names and remove trialing or leading whitespaces
    df[col] = df[col].str.strip().str.lower()


fix_str_col(df_scraped_listings_raw, col = 'city')

In [246]:
fix_str_col(df_scraped_listings_raw, col = 'city')

In [247]:
cities = sorted(df_scraped_listings_raw['city'].unique())
cities

['??',
 'anthem',
 'apache junction',
 'avondale',
 'black canyon city',
 'buckeye',
 'carefree',
 'casa grande',
 'cave creek',
 'chandler',
 'coolidge',
 'downtown phoenix',
 'el mirage',
 'flagstaff',
 'florence',
 'fort mcdowell',
 'fountain hills',
 'gilbert',
 'glendale',
 'gold canyon',
 'goodyear',
 'laveen',
 'litchfield park',
 'maricopa',
 'maricopa county',
 'mesa',
 'morristown',
 'new river',
 'old town scottsdale',
 'paradise valley',
 'peoria',
 'phoenix',
 'queen creek',
 'rio verde',
 'san tan valley',
 'scottsdale',
 'sun city',
 'sun city west',
 'sun lakes',
 'surprise',
 'tempe',
 'tolleson',
 'waddell',
 'wittmann',
 'youngtown']

In [248]:
len(cities)

45

note: we removed 7 categories of cities

#### Exploring Longitude & Latitude Column

In [135]:
latitudes = df_scraped_listings_raw['lat'].unique()
print(len(latitudes))

8445


note: there are latitudes & longitudes that vary up to the 9th decimal place. Latitude and longitude, however, might be better represented by the city, state, country, and zip code as a feature. We'll keep it for now and treat it as a continous feature in the dataframe. 

In [23]:
df_scraped_listings_raw.describe()

scraping_id          lon          lat     capacity     has_pool  \
count  8.445000e+03  8445.000000  8445.000000  8445.000000  8445.000000   
mean   1.022693e+07  -111.967109    33.510719     6.053641     0.730847   
std    1.086948e+07     0.167902     0.133915     3.282324     0.443546   
min    2.000000e+00  -112.712487    32.840029     1.000000     0.000000   
25%    5.753000e+03  -112.052714    33.450837     4.000000     0.000000   
50%    1.487100e+04  -111.935227    33.501307     6.000000     1.000000   
75%    2.188136e+07  -111.897167    33.601196     8.000000     1.000000   
max    2.470754e+07  -111.367664    34.076572    26.000000     1.000000   

       is_superhost  
count   8445.000000  
mean       0.241445  
std        0.427985  
min        0.000000  
25%        0.000000  
50%        0.000000  
75%        0.000000  
max        1.000000

In [78]:
len(df_scraped_listings_raw['scraping_id'].unique())

8445

#### Exploring Name in the Listing dataframe

In [251]:
df_scraped_listings_raw['name'].unique()

array(['Old Town Scottsdale Vacation Home & GIANTS Rental',
       'Upscale Townhouse South Scottsdale',
       'North Scottsdale/ PV Private Guest House w/pool', ...,
       'The Adobe Guest House "By the Kokes"-Historic Phx',
       'Super clean Power ranch home', '*THE lovely RED*'], dtype=object)

Note: interesting to note is that there are 8445 unique scrapping ID names, but only 8296 unique names for the listings. Which ones are duplicated?

In [253]:
unique_names = list(df_scraped_listings_raw['name'].unique())

In [254]:
df_scraped_listings_raw[df_scraped_listings_raw['name'].duplicated() == True]

scraping_id                                listing             city  \
776          1231  https://www.airbnb.com/rooms/14672894           peoria   
935          1435  https://www.airbnb.com/rooms/16380680       scottsdale   
1076         1621  https://www.airbnb.com/rooms/17359373       scottsdale   
1313         2442   https://www.airbnb.com/rooms/1900408       scottsdale   
1478         3232   https://www.airbnb.com/rooms/4006488       scottsdale   
1650         4024   https://www.airbnb.com/rooms/4787650       scottsdale   
1761         4483   https://www.airbnb.com/rooms/4931802       scottsdale   
1882         4936   https://www.airbnb.com/rooms/5040298          phoenix   
1983         5291   https://www.airbnb.com/rooms/5084735          phoenix   
2043         5522   https://www.airbnb.com/rooms/5105440       scottsdale   
2075         5696   https://www.airbnb.com/rooms/5134935          phoenix   
2081         5706   https://www.airbnb.com/rooms/5142044          phoenix   
2108         5748   https://www.airbnb.com/rooms/5142984           peoria   
2152         5911   https://www.airbnb.com/rooms/5170803          phoenix   
2186         6065   https://www.airbnb.com/rooms/5170230           peoria   
2193         6096   https://www.airbnb.com/rooms/5203433         glendale   
2194         6103   https://www.airbnb.com/rooms/5203040          phoenix   
2363         7063   https://www.airbnb.com/rooms/7331134       scottsdale   
2390         7265   https://www.airbnb.com/rooms/7686141       scottsdale   
2497         7777   https://www.airbnb.com/rooms/8613982       scottsdale   
2600         8209   https://www.airbnb.com/rooms/9407594          phoenix   
2641         8495   https://www.airbnb.com/rooms/9685406         chandler   
2651         8551   https://www.airbnb.com/rooms/9758290       scottsdale   
2678         8646   https://www.airbnb.com/rooms/9928220          phoenix   
2756         9007  https://www.airbnb.com/rooms/10354156             mesa   
2776         9129  https://www.airbnb.com/rooms/10453004       scottsdale   
2783         9150  https://www.airbnb.com/rooms/10547884       scottsdale   
2787         9156  https://www.airbnb.com/rooms/10573048       scottsdale   
2828         9301  https://www.airbnb.com/rooms/10771531       scottsdale   
2855         9387  https://www.airbnb.com/rooms/10909532           peoria   
...           ...                                    ...              ...   
7425     23061578  https://www.airbnb.com/rooms/23064425      queen creek   
7489     23151757  https://www.airbnb.com/rooms/23154667       scottsdale   
7501     23164625  https://www.airbnb.com/rooms/23167544       scottsdale   
7529     23193996  https://www.airbnb.com/rooms/23197034         chandler   
7639     23292456  https://www.airbnb.com/rooms/23295297   fountain hills   
7676     23326871  https://www.airbnb.com/rooms/23329617          gilbert   
7691     23359067  https://www.airbnb.com/rooms/23362078          phoenix   
7702     23387724  https://www.airbnb.com/rooms/23390519       scottsdale   
7704     23389880  https://www.airbnb.com/rooms/23392840       scottsdale   
7725     23414856  https://www.airbnb.com/rooms/23417680             mesa   
7769     23481461  https://www.airbnb.com/rooms/23484320       scottsdale   
7803     23505698  https://www.airbnb.com/rooms/23508508       scottsdale   
7898     23644200  https://www.airbnb.com/rooms/23646984          phoenix   
7923     23679474  https://www.airbnb.com/rooms/23682469       scottsdale   
7924     23679862  https://www.airbnb.com/rooms/23682713             mesa   
7934     23685065  https://www.airbnb.com/rooms/23687841          phoenix   
7983     23776074  https://www.airbnb.com/rooms/23779095       scottsdale   
7987     23776327  https://www.airbnb.com/rooms/23779145       scottsdale   
7989     23776346  https://www.airbnb.com/rooms/23779101       scottsdale   
7991     23776419  https://www.airbnb.com/rooms/23779096       scottsdale   
80

Note: though some of the names are duplicated, their property features are different, so we'll keep them in our analysis. This might provide insight into why or why not this feature would be useful in predicting price. Name tends to be listed by the host, and sometimes it might add superflurous words. This column is a candidate for removal since most of the description in the name is added in the other column features. We could, in theory, lemmatize this column and add X number of columns that correspond to certain words in the description. We could create a one-hot embedding for these words and use it as a potential feature in price prediction. 

#### Fixing bedroom, bathroom, cleaning fee columns

Problem: Pandas is reading it as a object column. There might be errors in some of the entries, so we'll find and fix that.